sound/soc/sof/imx/imx8m.c

```diff
@@ -0,0 +1,279 @@
+// SPDX-License-Identifier: (GPL-2.0 OR BSD-3-Clause)
+//
+// Copyright 2020 NXP
+//
+// Author: Daniel Baluta <daniel.baluta@nxp.com>
+//
+// Hardware interface for audio DSP on i.MX8M
+
+#include <linux/firmware.h>
+#include <linux/of_platform.h>
+#include <linux/of_address.h>
+#include <linux/of_irq.h>
+
+#include <linux/module.h>
+#include <sound/sof.h>
+#include <sound/sof/xtensa.h>
+#include <linux/firmware/imx/dsp.h>
+
+#include "../ops.h"
+
+#define MBOX_OFFSET	0x800000
+#define MBOX_SIZE	0x1000
+
+struct imx8m_priv {
+	struct device *dev;
+	struct snd_sof_dev *sdev;
+
+	/* DSP IPC handler */
+	struct imx_dsp_ipc *dsp_ipc;
+	struct platform_device *ipc_dev;
+};
+
+static void imx8m_get_reply(struct snd_sof_dev *sdev)
+{
+	struct snd_sof_ipc_msg *msg = sdev->msg;
+	struct sof_ipc_reply reply;
+	int ret = 0;
+
+	if (!msg) {
+		dev_warn(sdev->dev, "unexpected ipc interrupt\n");
+		return;
+	}
+
+	/* get reply */
+	sof_mailbox_read(sdev, sdev->host_box.offset, &reply, sizeof(reply));
+
+	if (reply.error < 0) {
+		memcpy(msg->reply_data, &reply, sizeof(reply));
+		ret = reply.error;
+	} else {
+		/* reply has correct size? */
+		if (reply.hdr.size != msg->reply_size) {
+			dev_err(sdev->dev, "error: reply expected %zu got %u bytes\n",
+				msg->reply_size, reply.hdr.size);
+			ret = -EINVAL;
+		}
+
+		/* read the message */
+		if (msg->reply_size > 0)
+			sof_mailbox_read(sdev, sdev->host_box.offset,
+					 msg->reply_data, msg->reply_size);
+	}
+
+	msg->reply_error = ret;
+}
+
+static int imx8m_get_mailbox_offset(struct snd_sof_dev *sdev)
+{
+	return MBOX_OFFSET;
+}
+
+static int imx8m_get_window_offset(struct snd_sof_dev *sdev, u32 id)
+{
+	return MBOX_OFFSET;
+}
+
+static void imx8m_dsp_handle_reply(struct imx_dsp_ipc *ipc)
+{
+	struct imx8m_priv *priv = imx_dsp_get_data(ipc);
+	unsigned long flags;
+
+	spin_lock_irqsave(&priv->sdev->ipc_lock, flags);
+	imx8m_get_reply(priv->sdev);
+	snd_sof_ipc_reply(priv->sdev, 0);
+	spin_unlock_irqrestore(&priv->sdev->ipc_lock, flags);
```

Comment:
out of curiosity, do you handle this lock in interrupt context?

Reply:
@plbossart yes, this is a stacktrace:

```
[   11.247874] Hardware name: FSL i.MX8MP EVK (DT)
[   11.252402] Call trace:
[   11.254853]  dump_backtrace+0x0/0x158
[   11.258514]  show_stack+0x14/0x20
[   11.261830]  dump_stack+0xbc/0x100
[   11.265234]  imx8_dsp_handle_reply+0x48/0x178 [snd_sof_imx8m]
[   11.270980]  imx_dsp_handle_rx+0x50/0x60
[   11.274903]  mbox_chan_received_data+0x18/0x28
[   11.279346]  imx_mu_isr+0x1b4/0x248
[   11.282836]  __handle_irq_event_percpu+0x70/0x198
[   11.287540]  handle_irq_event_percpu+0x34/0x88
[   11.291983]  handle_irq_event+0x44/0xc8
[   11.295818]  handle_fasteoi_irq+0xac/0x158
[   11.299914]  generic_handle_irq+0x24/0x38
[   11.303923]  __handle_domain_irq+0x5c/0xb8
[   11.308018]  gic_handle_irq+0x5c/0x150
[   11.311765]  el1_irq+0xb8/0x180
[   11.314907]  cpuidle_enter_state+0x138/0x340
[   11.319176]  cpuidle_enter+0x38/0x50
[   11.322751]  call_cpuidle+0x1c/0x48
[   11.326238]  do_idle+0x218/0x2b8
[   11.329465]  cpu_startup_entry+0x24/0x40
[   11.333388]  rest_init+0xd4/0xe0
[   11.336616]  arch_call_rest_init+0xc/0x14
[   11.340625]  start_kernel+0x42c/0x458
```